server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -269,8 +272,24 @@ static boolean executeBulkItemRequest(BulkPrimaryExecutionContext context, Updat
                 request.ifSeqNo(), request.ifPrimaryTerm());
         } else {
             final IndexRequest request = context.getRequestToExecute();
+            String id = request.id();
+            IndexSettings indexSettings = context.getPrimary().indexSettings();
+            if (indexSettings != null) {
+                int maxDocIdLen = context.getPrimary().indexSettings().getMaxDocIdLength();
+                if (id != null &&
+                    id.getBytes(StandardCharsets.UTF_8).length > maxDocIdLen) {
+                    ActionRequestValidationException validationException = new ActionRequestValidationException();
+                    validationException.addValidationError("id is too long, must be no longer than "
+                        + maxDocIdLen + " bytes but was: " +
+                        id.getBytes(StandardCharsets.UTF_8).length);
+                    result = new Engine.IndexResult(validationException, request.version());
+                    context.markOperationAsExecuted(result);
+                    context.markAsCompleted(context.getExecutionResult());
```

Comment:
This moves the error handling from request to item, meaning previously the entire bulk request would fail, whereas now, only the individual item will fail. I am not really opposed to this, but will need to double check whether we need backwards compatibility semantics on this change.

Likewise, this will also fail for update and delete I think? I would prefer to leave those out of this for now, since update/delete of a too long id will now error out instead of just being treated as "not found".